cmds/tail/tail.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Tail prints the lasts 10 lines of a file. Can additionally follow the
+// the end of the file as it grows.
+//
+// Synopsis:
+//     tail [-f] [-n lines_to_show] [FILE]
+//
+// Description:
+//     If no files are specified, read from stdin.
+//
+// Options:
+//     -f: follow the end of the file as it grows
+//     -n: specify the number of lines to show (default: 10)
+
+// Missing features:
+// - follow-mode (i.e. tail -f)
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"syscall"
+)
+
+var (
+	flagFollow   = flag.Bool("f", false, "follow the end of the file")
+	flagNumLines = flag.Int("n", 10, "specify the number of lines to show")
+)
+
+// TailConfig is a configuration object for the Tail function
+type TailConfig struct {
+	follow   bool // enable follow-mode (-f)
+	numLines int  // specifies the number of lines to print (-n)
+}
+
+// getBlockSize returns the number of bytes to read for each ReadAt call. This
+// helps minimize the number of syscalls to get the last N lines of the file.
+func getBlockSize(numLines int) int64 {
+	// This is currently computed as 81 * N, where N is the requested number of
+	// lines, and 81 is a relatively generous estimation of the average line
+	// length.
+	return 81 * int64(numLines)
+}
+
+// readLastLinesBackwards reads the last N lines from the provided file, reading
+// backwards from the end of the file. This is more efficient than reading from
+// the beginning, but can only be done on seekable files, (e.g. this won't work
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines int) error {
+	blkSize := getBlockSize(numLines)
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
```

Comment:
I think this invariant of always being at the end of the file makes this function a little complicated. if you just pass back the lastPos to the caller, the caller can decide to use it to seek (or ReadAt, or whatever) when -f is passed.

Reply:
I agree. In fact my original implementation was returning `(int64, error)` instead of just `error`. I changed that because the function `readLastLinesFromBeginning` cannot return a position. In C I would have used `ftell` but Go does not expose it. Searching on Google it seems that the common workaround is to `Seek(0, os.SEEK_CUR)`, which cannot be done in `readLastLinesFromBeginning`, as it's used for non-seekable files.

If I don't enforce this behaviour, I have to make the caller (`Tail()`) more cumbersome, because it will call the two functions that do the same thing (yet in different ways) with two different signatures, i.e.:

```
err = readLastLinesBackwards(inFile, writer, config.numLines)
...
if retryFromBeginning {
    err = readLastLinesFromBeginning(inFile, writer, config.numLines) 
}
```

I don't have strong opinions on whether it's better to have two discrepant signatures or to use a `io.ReaderAt` instead of `os.File`. Which is why it's good to discuss it here :)